workflow/executor/executor.go

```diff
@@ -931,6 +958,100 @@ func (we *WorkflowExecutor) Wait(ctx context.Context) error {
 	return nil
 }
 
+var progressRegexp = regexp.MustCompile("#argo .*progress=([0-9]+/[0-9]+)")
+
+// monitorProgress monitors for self-reported progress in the progressFile and patches the pod annotations with the parsed progress.
+//
+// The function reads the last line of the `progressFile` every `readFileTickDuration`.
+// If the line matches `#argo progress=N/M`, will set the progress annotation to the parsed progress value.
+// Every `annotationPatchTickDuration` the pod is patched with the updated annotations. This way the controller
+// gets notified of new self reported progress.
+func (we *WorkflowExecutor) monitorProgress(ctx context.Context, progressFile string) {
+	annotationPatchTicker := time.NewTicker(we.annotationPatchTickDuration)
+	defer annotationPatchTicker.Stop()
+	fileTicker := time.NewTicker(we.readProgressFileTickDuration)
+	defer fileTicker.Stop()
+
+	lastLine := ""
+	progressFile = filepath.Clean(progressFile)
+	// if progress has not been detected once within 10 ticks, abort monitoring
+	countDown := 10
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.WithError(ctx.Err()).Info("context done")
+			return
+		case <-annotationPatchTicker.C:
+			log.WithField("annotations", we.annotations).Infof("patching pod annotations")
+			err := we.patchPodAnnotations(ctx)
+			if err != nil {
+				log.WithError(err).Warn("failed to patch progress annotations")
+			}
+		case <-fileTicker.C:
+			if !we.isRunning(ctx) {
+				log.Info("pod not running yet, skipping tick")
+				continue
+			}
+
+			if countDown == 0 {
+				log.Info("progress not appeared after 10 ticks, cancelling monitoring")
+				return
+			}
+
+			data, err := ioutil.ReadFile(progressFile)
+			if err != nil {
+				log.WithError(err).WithField("file", progressFile).Info("unable to watch file")
+				if countDown > 0 {
+					countDown--
+				}
+				continue
+			}
+			lines := strings.Split(strings.TrimSpace(string(data)), "\n")
+			mostRecent := strings.TrimSpace(lines[len(lines)-1])
+
+			if mostRecent == "" || mostRecent == lastLine {
+				if countDown > 0 {
+					countDown--
+				}
+				continue
+			}
+			lastLine = mostRecent
+
+			v := progressRegexp.FindStringSubmatch(lastLine)
+			if len(v) == 2 {
+				progress, ok := wfv1.ParseProgress(v[1])
+				if ok {
+					log.WithField("progress", progress).Info()
+					we.annotations[common.AnnotationKeyProgress] = string(progress)
+					countDown = -1
+				} else {
+					log.WithField("v", v[1]).Info("unable to parse progress")
+					if countDown > 0 {
+						countDown--
+					}
+				}
+			}
+		}
+	}
+}
+
+func (we *WorkflowExecutor) isRunning(ctx context.Context) bool {
+	pod, err := we.getPod(ctx)
```

Comment:
this can be a very expensive API call, how can we avoid it?

Reply:
true. I haven't yet seen a way to check if the current pod is running or not, but I'm also not sure if it's _really_ needed. 

Without this check, the first few ticks of monitorProgress will result in trying to read a file which does not yet exist. 
This can lead to two potential issues:
1. many info logs about the file not existing until the pod creates it (could be solved by ignoring file not exists errors)
2. together with the `countDown` feature to abort progress monitoring early, it may lead to aborting it too early, i.e. when the pod hasn't had a chance to run yet

if we solve those two issues, I think we can get rid of it.

Reply:
looking at the code, we should improve it anyway, the check if pod is running is only required until the pod is running (obvious, isn't it? ;)). Currently it checks it on every read file tick, which is too much.

Reply:
>  ignoring file not exists 

yes, ignore this error

>  running is only required until the pod is running 

I think you can only check if the pod is running, if the pod is running. So will always be true?